rust/src/witness_builder.rs

```diff
@@ -0,0 +1,389 @@
+                0x81, 0x2a, 0xb2, 0xfd, 0xa5, 0x78, 0x60, 0xac, 0x87, 0x6b, 0xc4, 0x89, 0x19, 0x2c, 0x1e, 0xf4, 0xce, 0x25, 0x3c, 0x19, 0x7e, 0xe2, 0x19, 0xa4]
+        ).unwrap()
+    }
+
+    fn fake_private_key2() -> Bip32PrivateKey {
+        Bip32PrivateKey::from_bytes(
+            &hex::decode("d84c65426109a36edda5375ea67f1b738e1dacf8629f2bb5a2b0b20f3cd5075873bf5cdfa7e533482677219ac7d639e30a38e2e645ea9140855f44ff09e60c52c8b95d0d35fe75a70f9f5633a3e2439b2994b9e2bc851c49e9f91d1a5dcbb1a3").unwrap()
+        ).unwrap()
+    }
+    
+
+    #[test]
+    fn vkey_test() {
+        let mut builder = TransactionWitnessSetBuilder::new();
+        
+        let raw_key_public = fake_raw_key_public(0);
+        let fake_sig = fake_raw_key_sig(0);
+
+        // add the same element twice
+        builder.add_vkey(&Vkeywitness::new(
+            &Vkey::new(&raw_key_public),
+            &fake_sig
+        ));
+        builder.add_vkey(&Vkeywitness::new(
+            &Vkey::new(&raw_key_public),
+            &fake_sig
+        ));
+
+        // add a different element
+        builder.add_vkey(&Vkeywitness::new(
+            &Vkey::new(&fake_raw_key_public(1)),
+            &fake_raw_key_sig(1)
+        ));
+
+        let wit_set = builder.build().unwrap();
+        assert_eq!(
+            wit_set.vkeys().unwrap().len(),
+            2
+        );
+    }
+
+    #[test]
+    fn bootstrap_test() {
+        let mut builder = TransactionWitnessSetBuilder::new();
+
+        // add the same element twice
+        let wit1 = make_icarus_bootstrap_witness(
+            &TransactionHash::from([0u8; TransactionHash::BYTE_COUNT]),
+            &ByronAddress::from_base58("Ae2tdPwUPEZGUEsuMAhvDcy94LKsZxDjCbgaiBBMgYpR8sKf96xJmit7Eho").unwrap(),
+            &fake_private_key1()
+        );
+        builder.add_bootstrap(&wit1);
+        builder.add_bootstrap(&wit1);
+
+        // add a different element
+        builder.add_bootstrap(&make_icarus_bootstrap_witness(
+            &TransactionHash::from([0u8; TransactionHash::BYTE_COUNT]),
+            &ByronAddress::from_base58("Ae2tdPwUPEZGUEsuMAhvDcy94LKsZxDjCbgaiBBMgYpR8sKf96xJmit7Eho").unwrap(),
+            &fake_private_key2()
+        ));
+
+        let wit_set = builder.build().unwrap();
+        assert_eq!(
+            wit_set.bootstraps().unwrap().len(),
+            2
+        );
+    }
+
+    #[test]
+    fn native_script_test() {
+        let mut builder = TransactionWitnessSetBuilder::new();
+
+        // add the same element twice
+        let wit1 = NativeScript::new_timelock_start(
+            &TimelockStart::new(1),
+        );
+        builder.add_native_script(&wit1);
+        builder.add_native_script(&wit1);
+
+        // add a different element
+        builder.add_native_script(&NativeScript::new_timelock_start(
+            &TimelockStart::new(2),
+        ));
+
+        let wit_set = builder.build().unwrap();
```

Comment:
We have the not-enough-errors case tested in `requirement_test()` for failure but not for success anywhere. All of these tests here were written before `RequiredWitnessSet` was added so they don't use it. We should modify at least one of these tests (or all or write a separate one) to include a `RequiredWitnessSet` to test that path too as these will all succeed since there are no required things.

Reply:
We have  `requirement_test` for testing the `RequiredWitnessSet` but I added a `requirement_test_pass` that makes sure the success case works for everything